abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,1471 @@
+  - `i32 (size_t) buffer_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Sets content of the buffer `buffer_type` to the values from memory slices
+pointed by the provided list of vectors (`buffer_data`, `buffer_size`),
+replacing `size` bytes in the existing buffer, starting at the `offset`. Host
+environment must adjust the size of the destination buffer (e.g. shrink the
+buffer if target `size` is larger than `buffer_size`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for unknown `buffer_type`.
+
+Returns `InvalidMemoryAccess` when `buffer_data` and/or `buffer_size` point to
+invalid memory address.
+
+#### `proxy_get_map_values`
+
+* params:
+  - `i32 (proxy_map_type_t) map_type`
+  - `i32 (iovec_array_t) keys_data`
+  - `i32 (size_t) keys_size`
+  - `i32 (iovec_array_t[]*) return_map_data`
+  - `i32 (size_t*) return_map_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Returns map containing key-values for keys from the provided list of vectors
+(`keys_data`, `keys_size`) from the map (`map_type`) in a list of lists
+of vectors (`return_map_data`, `return_map_size`) in which the first element
+in each list represents a key, followed by a zero or more values.
+
+If the list of keys is empty, then all key-values in the map are returned.
+
+In context of HTTP header maps, the special key `"*"` can be used to retrieve
+all regular HTTP headers, and the special key `":*"` can be used to retrieve
+all pseudo-headers.
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for unknown `map_type`.
+
+Returns `InvalidMemoryAccess` when `key_data`, `key_size`, `return_values_data`
+and/or `return_values_size` point to invalid memory address.
+
+#### `proxy_set_map_values`
+
+* params:
+  - `i32 (proxy_map_type_t) map_type`
+  - `i32 (iovec_array_t) remove_keys_data`
+  - `i32 (size_t) remove_keys_size`
+  - `i32 (iovec_array_t[]) map_data`
+  - `i32 (size_t) map_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Adds key-values from the provided list of lists of vectors (`map_data`,
+`map_size`) to the given map (`map_type`). The first element in each list
+represents a key, followed by a zero or more values. For the existing keys,
+new values will be added to the existing ones.
+
+Keys to remove and/or which values will be replaced can be provided in an
+optional list of vectors (`remove_keys_data`, `remove_keys_size`).
+
+In context of HTTP header maps, the special key `"*"` can be used to remove
+all regular HTTP headers, and the special key `":*"` can be used to remove
+all pseudo-headers.
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for unknown `map_type`.
+
+Returns `InvalidMemoryAccess` when `key_data`, `key_size`, `values_data` and/or
+`values_size` point to invalid memory address.
+
+
+## Shared key-value store
+
+Note: Global key-value store can be always accessed using `kvstore_id` `0`.
+
+### Functions exposed by the host environment
+
+#### `proxy_open_shared_kvstore`
```

Comment:
Why can't the kvstore be used to model headers? It seems a bit awkward to have two different ABI mechanisms for key-values.